### PR TITLE
Add kauth kernel extension.

### PR DESCRIPTION
**Pull request by babakbahamin**

### CODE REVIEW

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

**Comment by marpaia** [comment]
image

**Comment by marpaia** [comment]
@theopolis and I were chatting about this yesterday, trying to figure out how we're going to handle merging this together with what @mofarrell is working on in #1196. I believe that we left off feeling that we should merge this and then iterate on adding xplatform communication functionality, additional capabilities, etc.

Does anyone have any objections of merging this as is and then submitting additional pull requests to improve and iterate on the capabilities?

**Comment by theopolis** [comment]
Agreed, this was @mofarrell's suggestion. If he feels comfortable iterating off of this base let's merge.

File: kernel/darwin/iokit/driver.cpp [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+  kernel_osquery_message_t message = {0};
+  clock_get_calendar_microtime(
+      &(message.time_secs), &(message.time_microsecs));
+  message.action = action;
+  message.pid = proc_selfpid();
+  message.ppid = proc_selfppid();
+  message.uid = kauth_cred_getuid(credential);
+  message.ruid = kauth_cred_getruid(credential);
+  proc_selfname(message.proc_name, MAXCOMLEN + 1);
+  if (arg0)
```
Comment (theopolis) [inline kernel/darwin/iokit/driver.cpp]: nit, can you wrap these in `{}`?

File: kernel/darwin/iokit/driver.cpp [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+#include <kern/clock.h>
+#include <sys/proc.h>
+#include <sys/vnode.h>
+
+#include "kernel/darwin/iokit/public.h"
+#include "kernel/darwin/iokit/logging.h"
+
+// This required macro defines the class's constructors, destructors,
+// and several other methods I/O Kit requires.
+OSDefineMetaClassAndStructors(com_facebook_iokit_OSQuery, IOService)
```
Comment (theopolis) [inline kernel/darwin/iokit/driver.cpp]: nit, we should use `com_facebook_iokit_osquery`. No caps unless osquery is starting a sentence or method name, in which case it'll be `Osquery`.

File: kernel/darwin/iokit/driver.cpp [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+bool OSQueryDriverClassName::start(IOService *provider) {
+  bool result = super::start(provider);
+  LOGD("Starting.");
+  // Create and initialize an IODataQueue object.
+  data_queue_ = IOSharedDataQueue::withCapacity(
+      kQueueEventCapacity
+      * (sizeof(kernel_osquery_message_t) + DATA_QUEUE_ENTRY_HEADER_SIZE));
+  if (!data_queue_) {
```
Comment (theopolis) [inline kernel/darwin/iokit/driver.cpp]: I'm a big fan on explicit comparisons to `nullptr` when appropriate. But definitely up to you.

File: kernel/tools/darwin/build/osquery.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+      productType = "com.apple.product-type.kernel-extension";
+    };
+/* End PBXNativeTarget section */
+
+/* Begin PBXProject section */
+    A744FD2E1B2B8011009A5085 /* Project object */ = {
+      isa = PBXProject;
+      attributes = {
+        LastUpgradeCheck = 0610;
+        ORGANIZATIONNAME = "Babak Bahamin";
```
Comment (theopolis) [inline kernel/tools/darwin/build/osquery.xcodeproj/project.pbxproj]: You could ditch this.

File: kernel/darwin/iokit/user_client.cpp [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+IOReturn OSQueryUserClientClassName::clientMemoryForType(
+    UInt32 type, IOOptionBits *options, IOMemoryDescriptor **memory) {
+  *memory = nullptr;
+  *options = 0;
+  if (type != kIODefaultMemoryType)
+    return kIOReturnNoMemory;
+
+  // You don't need to balance with a release() because the user client does
+  // the release() for you.
+  IOMemoryDescriptor *memory_to_share = provider_->GetSharedMemory();
```
Comment (theopolis) [inline kernel/darwin/iokit/user_client.cpp]: Is it not possible to retrieve and retain this as const? The user client should not be communicating back by manipulating the struct? Maybe up the retain count within the `GetSharedMemory` implementation?

Comment (babakbahamin) [inline kernel/darwin/iokit/user_client.cpp]: I am not sure I understand. Are you saying you want the retain() to be done in GetSharedMemory() versus here?

Comment (theopolis) [inline kernel/darwin/iokit/user_client.cpp]: I don't know if there are second-order effects from moving it into `GetSharedMemory()`. If you can, then you can return a const descriptor to the user, which would be ideal.

**Pull request merged**